### PR TITLE
Add kairosdb Sink

### PR DESCRIPTION
**Pull request by psycho-ir**
KairosDB Sink is here 🍨 
Will do the documentation after getting feedback.

### CODE REVIEW

**Comment by 2m** [comment]
This looks good! Thanks for bringing one more connector to the Alpakka family. :) Currently I do not have time to review this in more detail. Ping me if no one from the community has not reviewed this by the next week.

File: kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+private[kairosdb] class KairosSinkStageLogic(
+    in: Inlet[MetricBuilder],
+    shape: SinkShape[MetricBuilder],
+    kairosClient: Client,
+    promise: Promise[Done],
+    settings: KairosSinkSettings
+)(executionContext: ExecutionContext)
+    extends GraphStageLogic(shape)
+    with StageLogging {
+
```
Comment (francisdb) [inline kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala]: Would it not be better to force an explicit EC as this uses blocking code?

File: kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/javadsl/KairoSink.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import akka.stream.alpakka.kairosdb.{KairosDBSinkStage, KairosSinkSettings, NullExecutionContext}
+import akka.stream.javadsl.Sink
+import org.kairosdb.client.Client
+import org.kairosdb.client.builder.MetricBuilder
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+ * Created by SOROOSH on 3/22/17.
+ */
```
Comment (francisdb) [inline kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/javadsl/KairoSink.scala]: This doc should probably be removed / replaced with something more useful?

**Comment by francisdb** [comment]
Might take some time later to test this on our staging kairos

File: kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala (hunk 1 above)
Comment (psycho-ir) [inline kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala]: Actually in the first implementation I did it in this way. 
But later on decided to keep it similar to what already is there in alpakka. e.g. IranMQ implementation
https://github.com/akka/alpakka/blob/master/ironmq/src/main/scala/akka/stream/alpakka/ironmq/IronMqPullStage.scala#L48
But still is still more flexible as client can pass different EC as well.

File: kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/javadsl/KairoSink.scala (hunk 2 above)
Comment (psycho-ir) [inline kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/javadsl/KairoSink.scala]: Yes of course :) 
I will document the API after we finalize the document.

**Comment by psycho-ir** [comment]
@francisdb No I haven't tried this version yet on the production. but in staging works well.

File: kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala (hunk 1 above)
Comment (francisdb) [inline kairosdb/src/main/scala/akka/stream/alpakka/kairosdb/KairosDBSinkStage.scala]: _Let me first fix my wording. I mean removing the **default** value for the implicit ec but keeping it implicit._

The IronMQ client is async using callbacks, there is no blocking{...} in there so this should not be an issue. I wonder if the internal ec from akka streams is set up to grow when there are blocking calls (like the global one).
Anyway, when implicit without default the user can still decide to provide mat.executionContext.

```scala
class KairosDBSinkStage(settings: KairosSinkSettings, kairosClient: Client)(implicit executionContext: ExecutionContext)
```

**Comment by francisdb** [comment]
Could even be better to just write a client using akka http instead of using the blocking java client. Copying some stuff from https://github.com/waylayio/kairosdb-scala which would avoid the play dependencies.

**Comment by psycho-ir** [comment]
@francisdb 
executionContext has no default valye anymore.

**Comment by 2m** [comment]
I have taken a closer took and now wonder if this could be implemented without a graph stage. Since you are wrapping calls to the kairos db in a future, then it could be implemented as

```scala
Source[MetricBuilder](...).mapAsync(parallelism, Future { kairosClient.pushMetrics(builder) })
```

Since the client is blocking this would still require a more careful management of execution contexts.

I also liked @francisdb suggestion to reimplement kairosdb client using akka-http. For starters it could be only the `pushMetrics` operation and that one could also be implemented with `Http.singleRequest` (which takes a request and returns a future of response) and mapAsync.

Let me know what you thing.

**Comment by yaroot** [comment]
Hi, though I don't use kairosdb anymore but I'm use akka-stream for a lot of similar stuff. It could be implemented by just using flow.

```
    def submitDatapoints(...) = // call kairosdb http endpoint, just need a json library and use akka-http here

    case class Datapoint(name: String, datapoints: Seq[Seq[String]], tags: Seq[String], ttl: Long)

    Flow[Datapoint]
      .batch(100, Vector.newBuilder[Datapoint].+=)(_ += _) // batch http calls for efficiency
      .mapAsync(5) { datapoints =>
        submitDatapoints(datapoints.result())
      }
      .to(Sink.ignore)
```

And if you have to block you could use the built-in blocking pool via `Flow.addAttributes(Attributes(IODispatcher))`

just my 2 cents

**Comment by ennru** [comment]
This PR hasn't been touched for a very long time.
Please re-open or create a new to continue this work.

**Pull request closed without merging**